### PR TITLE
Adding support for ipfs large files

### LINKED ISSUE #8370: Implement ipfs for large files.

### PR DESCRIPTION
**Pull request by mijovic**
Fixes #8370

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ Compiler Features:
 Bugfixes:
  * isoltest: Added new keyword `wei` to express function value in semantic tests
  * Standard-JSON-Interface: Fix a bug related to empty filenames and imports.
+ * ipfs-hashing: Added support for large files that we need to split in multiple chunks.
```
Comment (leonardoalt) [inline Changelog.md]: I think this should rather do to the `Compiler Features` section

File: libsolutil/IpfsHash.cpp [hunk 2]
```diff
@@ -55,34 +55,124 @@ string base58Encode(bytes const& _data)
...
+		// filesize: length as varint
+		protobufEncodedData += bytes{0x18} + lengthAsVarint;
+
+		// PBDag:
+		// Data: (length delimited bytes)
+		size_t protobufLength = protobufEncodedData.size();
+		bytes blockData = bytes{0x0a} + varintEncoding(protobufLength) + std::move(protobufEncodedData);
+
+		// Multihash: sha2-256, 256 bits
+		chunkLevels[0].emplace_back(
```
Comment (leonardoalt) [inline libsolutil/IpfsHash.cpp]:
```suggestion
		chunkLevels.at(0).emplace_back(
```

File: libsolutil/IpfsHash.cpp [hunk 3]
```diff
@@ -55,34 +55,124 @@ string base58Encode(bytes const& _data)
...
+		blockSize(_blockSize)
+	{}
+	bytes hash;
+	size_t size;
+	size_t blockSize;
+};
+
+typedef vector<Chunk> Chunks;
+
+/// It build tree starting from bottom level where nodes are data nodes.
```
Comment (leonardoalt) [inline libsolutil/IpfsHash.cpp]:
```suggestion
/// Builds a tree starting from the bottom level where nodes are data nodes.
```

File: libsolutil/IpfsHash.cpp [hunk 4]
```diff
@@ -55,34 +55,124 @@ string base58Encode(bytes const& _data)
...
+/// Data nodes should be calculated and passed as the only level in chunk levels
+/// Each next level is calculated as following:
+///   - Pick up to maxChildNum (174) nodes until whole level is added, group them and pass to the node in next level
```
Comment (leonardoalt) [inline libsolutil/IpfsHash.cpp]:
```suggestion
///   - Pick up to maxChildNum (174) nodes until a whole level is added, group them and pass to the node in the next level
```

File: libsolutil/IpfsHash.cpp [hunk 5]
```diff
@@ -55,34 +55,124 @@ string base58Encode(bytes const& _data)
...
+///   - Do this until current level has only one node, that node stores hash for
```
Comment (leonardoalt) [inline libsolutil/IpfsHash.cpp]:
```suggestion
///   - Do this until the current level has only one node, return the hash in that node
```

File: libsolutil/IpfsHash.cpp [hunk 6]
```diff
@@ -55,34 +55,124 @@ string base58Encode(bytes const& _data)
...
+		bytes newLengths;
+		size_t newChunkSize = 0;
+		size_t newBlockSize = 0;
+
+		for (size_t chunkIndex = 0; chunkIndex < chunkLevels[currentLevel].size(); chunkIndex++)
+		{
+			// new link that we are adding
+			bytes newLink = bytes {0x0a} + varintEncoding(chunkLevels[currentLevel][chunkIndex].hash.size());
+			newLink += std::move(chunkLevels[currentLevel][chunkIndex].hash);
+			newLink += bytes{0x12,0x00,0x18} + varintEncoding(chunkLevels[currentLevel][chunkIndex].blockSize);
```
Comment (leonardoalt) [inline libsolutil/IpfsHash.cpp]:
```suggestion
			newLink += bytes{0x12, 0x00, 0x18} + varintEncoding(chunkLevels[currentLevel][chunkIndex].blockSize);
```

File: libsolutil/IpfsHash.cpp [hunk 7]
```diff
@@ -55,34 +55,124 @@ string base58Encode(bytes const& _data)
...
+				);
+
+				// we need to reset as we are about to build another node
+				newLinks = {};
+				newLengths = {};
+				newChunkSize = 0;
+				newBlockSize = 0;
+			}
+		}
+		chunkLevels.push_back(std::move(nextLevel));
```
Comment (leonardoalt) [inline libsolutil/IpfsHash.cpp]:
```suggestion
		chunkLevels.emplace_back(std::move(nextLevel));
```

**Review by leonardoalt — COMMENTED** [review]
Just an initial look

File: Changelog.md (hunk 1 above)
Comment (mijovic) [reply Changelog.md]: Moved

File: libsolutil/IpfsHash.cpp (hunk 2 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: maybe chunkLevels.front() instead?

File: libsolutil/IpfsHash.cpp (hunk 3 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: Done

File: libsolutil/IpfsHash.cpp (hunk 4 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: Done

File: libsolutil/IpfsHash.cpp (hunk 5 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: Done

File: libsolutil/IpfsHash.cpp (hunk 6 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: Fixed

File: libsolutil/IpfsHash.cpp (hunk 7 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: Fixed

File: libsolutil/IpfsHash.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolutil/IpfsHash.cpp]: Sounds good!

**Comment by chriseth** [comment]
Please don't mention issue numbers in commit messages, it creates unnecessary notifications.

**Comment by leonardoalt** [comment]
I think this PR should also remove `if (scanner->source().size() < 1024 * 256)` from `CompilerStack::ipfsUrl`, and a similar if from `CompilerStack::createCBORMetadata`.

**Comment by mijovic** [comment]
> I think this PR should also remove `if (scanner->source().size() < 1024 * 256)` from `CompilerStack::ipfsUrl`, and a similar if from `CompilerStack::createCBORMetadata`.

Thanks for pointing out. I updated PR with those two changes.

File: test/libsolutil/IpfsHash.cpp [hunk 8]
```diff
@@ -60,15 +60,23 @@ BOOST_AUTO_TEST_CASE(test_largest_unchunked)
...
-//	BOOST_REQUIRE_EQUAL(data.size(), length);
-//	BOOST_CHECK_EQUAL(ipfsHashBase58(data), "QmNg7BJo8gEMDK8yGQbHEwPtycesnE6FUULX5iVd5TAL9f");
-//}
+BOOST_AUTO_TEST_CASE(test_large)
+{
+	size_t length = 1310710;
+	string data;
+	data.resize(length, 0);
+	BOOST_REQUIRE_EQUAL(data.size(), length);
+	BOOST_CHECK_EQUAL(ipfsHashBase58(data), "QmNg7BJo8gEMDK8yGQbHEwPtycesnE6FUULX5iVd5TAL9f");
```
Comment (leonardoalt) [inline test/libsolutil/IpfsHash.cpp:78]: @mijovic did you check locally that these hashes are indeed correct?

Comment (mijovic) [reply test/libsolutil/IpfsHash.cpp:78]: Yes, I checked locally, as well as in https://dag.ipfs.io/ with correct settings.

File: libsolutil/IpfsHash.cpp [hunk 9]
```diff
@@ -55,34 +55,125 @@ string base58Encode(bytes const& _data)
...
+		hash(std::move(_hash)),
+		size(_size),
+		blockSize(_blockSize)
+	{}
+	bytes hash;
+	size_t size;
+	size_t blockSize;
+};
+
+typedef vector<Chunk> Chunks;
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]:
```suggestion
using Chunks = vector<Chunk>;
```

File: libsolutil/IpfsHash.cpp [hunk 10]
```diff
@@ -55,34 +55,125 @@ string base58Encode(bytes const& _data)
 }
 }
 
+struct Chunk
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp:72]: Why not put it inside the anonymous namespace?

File: libsolutil/IpfsHash.cpp [hunk 11]
```diff
@@ -55,34 +55,125 @@ string base58Encode(bytes const& _data)
...
+};
+
+typedef vector<Chunk> Chunks;
+
+/// Builds a tree starting from the bottom level where nodes are data nodes.
+/// Data nodes should be calculated and passed as the only level in chunk levels
+/// Each next level is calculated as following:
+///   - Pick up to maxChildNum (174) nodes until a whole level is added, group them and pass to the node in the next level
+///   - Do this until the current level has only one node, return the hash in that node
+bytes groupChunksBottomUp(vector<Chunks>& chunkLevels)
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]: Please move into anonymous namespace above.

File: libsolutil/IpfsHash.cpp [hunk 12]
```diff
@@ -55,34 +55,125 @@ string base58Encode(bytes const& _data)
...
-	// Type: File
-	protobufEncodedData += bytes{0x08, 0x02};
-	if (!_data.empty())
+	for (unsigned long chunkIndex = 0; chunkIndex < chunkCount; chunkIndex++)
 	{
-		// Data (length delimited bytes)
-		protobufEncodedData += bytes{0x12};
-		protobufEncodedData += lengthAsVarint;
-		protobufEncodedData += asBytes(std::move(_data));
+		std::string chunkData =
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]: Can you use a `bytesRef`? Then we would not have to copy the string.

Comment (chriseth) [reply libsolutil/IpfsHash.cpp]: Or acutally - we use it in `asBytes(...)` below - so maybe it makes sense to convert it to a bytes already here?

File: libsolutil/IpfsHash.cpp (hunk 9 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: Fixed

File: libsolutil/IpfsHash.cpp (hunk 10 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp:72]: Fixed

File: libsolutil/IpfsHash.cpp (hunk 12 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: Totally makes sense to convert to bytes earlier

**Comment by chriseth** [comment]
Can you add tests around the point where the size needs to be broken into multiple chunks and where another levels is added?

File: libsolutil/IpfsHash.cpp [hunk 13]
```diff
@@ -53,36 +53,127 @@ string base58Encode(bytes const& _data)
...
+};
+
+using Chunks = vector<Chunk>;
+
+/// Builds a tree starting from the bottom level where nodes are data nodes.
+/// Data nodes should be calculated and passed as the only level in chunk levels
+/// Each next level is calculated as following:
+///   - Pick up to maxChildNum (174) nodes until a whole level is added, group them and pass to the node in the next level
+///   - Do this until the current level has only one node, return the hash in that node
+bytes groupChunksBottomUp(vector<Chunks>& chunkLevels)
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]:
```suggestion
bytes groupChunksBottomUp(vector<Chunks>& _chunkLevels)
```

Maybe it even makes sense to even use `groupChunksBottomUp(Chunks _allChunks)`

and then do

```
vector<Chunks> chunkLevels;
chunkLevels.front() = std::move(_allChunks);
```

File: libsolutil/IpfsHash.cpp [hunk 14]
```diff
@@ -53,36 +53,127 @@ string base58Encode(bytes const& _data)
...
+	while (chunkLevels[currentLevel].size() != 1)
+	{
+		vector<Chunk> nextLevel;
+		Chunk newChunk;
+		bytes newLinks;
+		bytes newLengths;
+		size_t newChunkSize = 0;
+		size_t newBlockSize = 0;
+
+		for (size_t chunkIndex = 0; chunkIndex < chunkLevels[currentLevel].size(); chunkIndex++)
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]: This loop only uses `chunkLevels` as `chunkLevels[currentLevel]` - is it possible to extract it into its own function?

File: libsolutil/IpfsHash.cpp [hunk 15]
```diff
@@ -53,36 +53,127 @@ string base58Encode(bytes const& _data)
...
+///   - Do this until the current level has only one node, return the hash in that node
+bytes groupChunksBottomUp(vector<Chunks>& chunkLevels)
+{
+	size_t const maxChildNum = 174;
+	size_t currentLevel = 0;
+
+	// when we reach root it will be the only node in that level
+	while (chunkLevels[currentLevel].size() != 1)
+	{
+		vector<Chunk> nextLevel;
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]: Do you really need that many variables at this level?

**Comment by mijovic** [comment]
> Can you add tests around the point where the size needs to be broken into multiple chunks and where another levels is added?

I added one with 178 chunk and it creates 2 nodes in next level, than groups those two in final one.

Will add more of tests if needed.

File: libsolutil/IpfsHash.cpp (hunk 13 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: I am not sure if I follow this one. How would we call the function in this case?

Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: I see now

**Comment by leonardoalt** [comment]
@mijovic yes but can you also add tests that have
- just enough bytes to have another chunk (`nbytes(chunk)+1` bytes)
- `nbytes(chunk)` bytes

**Comment by mijovic** [comment]
> @mijovic yes but can you also add tests that have
> 
> * just enough bytes to have another chunk (`nbytes(chunk)+1` bytes)
> * `nbytes(chunk)` bytes

I added this one as well. Also, there is test with just enough to do multilevel (1024*256*174 + 1)

File: Changelog.md [hunk 16]
```diff
@@ -9,6 +9,7 @@ Language Features:
 Compiler Features:
  * AssemblyStack: Support for source locations (source mappings) and thus debugging Yul sources.
  * Commandline Interface: Enable output of experimental optimized IR via ``--ir-optimized``.
+ * ipfs-hashing: Added support for large files that we need to split in multiple chunks.
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
 * Metadata: Added support for IPFS hashes of large files that need to be split in multiple chunks.
```

Comment (chriseth) [reply Changelog.md]: Also, I think this needs to be moved to 0.6.5

File: libsolutil/IpfsHash.cpp [hunk 17]
```diff
@@ -53,36 +53,133 @@ string base58Encode(bytes const& _data)
...
+{
+	Chunk() = default;
+	Chunk(bytes _hash, size_t _size, size_t _blockSize):
+		hash(std::move(_hash)),
+		size(_size),
+		blockSize(_blockSize),
+		links{},
+		lengths{}
+	{}
+	bytes hash = {};
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp:81]: Please move members to the bottom

File: libsolutil/IpfsHash.cpp [hunk 18]
```diff
@@ -53,36 +53,133 @@ string base58Encode(bytes const& _data)
...
+
+		// PBDag:
+		// Data: (length delimited bytes)
+		size_t protobufLength = protobufEncodedData.size();
+		bytes blockData = bytes{0x0a} + varintEncoding(protobufLength) + std::move(protobufEncodedData);
+
+		// Multihash: sha2-256, 256 bits
+		allChunks.emplace_back(
+			bytes{0x12, 0x20} + picosha2::hash256(blockData),
+			chunkBytes.size() ,
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]:
```suggestion
			chunkBytes.size(),
```

File: libsolutil/IpfsHash.cpp [hunk 19]
```diff
@@ -53,36 +53,133 @@ string base58Encode(bytes const& _data)
...
+		size(_size),
+		blockSize(_blockSize),
+		links{},
+		lengths{}
+	{}
+	bytes hash = {};
+	size_t size = 0;
+	size_t blockSize = 0;
+
+	bytes links = {};
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]: Is it possible to separate blocks and links? Also the `updateHash` does not sound like a good idea.

File: libsolutil/IpfsHash.cpp [hunk 20]
```diff
@@ -53,36 +53,133 @@ string base58Encode(bytes const& _data)
...
+{
+	size_t const maxChildNum = 174;
+
+	Chunks nextLevel;
+	Chunk newChunk;
+
+	for (size_t chunkIndex = 0; chunkIndex < _currentLevel.size(); chunkIndex++)
+	{
+		// new link that we are adding
+		newChunk.addLink(_currentLevel[chunkIndex]);
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]: Can you collect the links in a vector locally here and then call a function that computes the link hash of the full vector instead?

Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: I will, it is more clean than current solution

File: libsolutil/IpfsHash.cpp [hunk 21]
```diff
@@ -53,36 +53,133 @@ string base58Encode(bytes const& _data)
...
+			bytes{0x12, 0x00, 0x18} + varintEncoding(_chunkLink.blockSize);
+
+		links += bytes{0x12} + varintEncoding(newLinkData.size()) + std::move(newLinkData);
+		lengths += bytes{0x20} + varintEncoding(_chunkLink.size);
+	}
+
+	void updateHash()
+	{
+		bytes tmp = bytes{0x08, 0x02, 0x18} + varintEncoding(size) + lengths;
+		bytes blockData = links + bytes{0x0a} + varintEncoding(tmp.size()) + tmp;
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]: Can you create a helper function `encodeByteArray(bytes const& _data) { return bytes{0x0a} + varintEncoding(_data.size()) + _data; }`?

File: libsolutil/IpfsHash.cpp (hunk 19 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: When you say blocks and links, you mean data chunks and link chunks?

**Comment by chriseth** [comment]
Re-checked all hashes in the test file and they are correct.

File: libsolutil/IpfsHash.cpp [hunk 22]
```diff
@@ -53,36 +53,133 @@ string base58Encode(bytes const& _data)
...
+
+	Chunks nextLevel;
+	Chunk newChunk;
+
+	for (size_t chunkIndex = 0; chunkIndex < _currentLevel.size(); chunkIndex++)
+	{
+		// new link that we are adding
+		newChunk.addLink(_currentLevel[chunkIndex]);
+
+		if (chunkIndex % maxChildNum == maxChildNum - 1 || chunkIndex == _currentLevel.size() - 1)
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]: If you collect the links in a local vector, you can use the vector's size for this check and do something like:
```
for (Chunk& chunk: _currentLevel)
{
  links.emplace_back(createLink(chunk));
  if (links.size() == maxChildNum - 1)
  {
    nextLevel.emplace_back(combineLinks(links));
    links = {};
  }
}
if (!links.empty())
    nextLevel.emplace_back(combineLinks(links));

return nextLevel;
}
```

File: libsolutil/IpfsHash.cpp (hunk 19 above)
Comment (chriseth) [reply libsolutil/IpfsHash.cpp]: I think so

File: libsolutil/IpfsHash.cpp (hunk 21 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: Created, added also one more reusable function to reduce code repetition.

File: libsolutil/IpfsHash.cpp (hunk 22 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: I created different structure to hold Link data and used local vector of links to store them

File: libsolutil/IpfsHash.cpp [hunk 23]
```diff
@@ -53,36 +68,142 @@ string base58Encode(bytes const& _data)
...
+
+	bytes hash = {};
+	size_t size = 0;
+	size_t blockSize = 0;
+};
+
+struct Link
+{
+	Link() = default;
+	Link(size_t _size, size_t _blockSize, bytes _hash) : size(_size), blockSize(_blockSize), hash(std::move(_hash)) {}
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]:
```suggestion
	Link(size_t _size, size_t _blockSize, bytes _hash): size(_size), blockSize(_blockSize), hash(std::move(_hash)) {}
```

File: libsolutil/IpfsHash.cpp [hunk 24]
```diff
@@ -53,36 +68,142 @@ string base58Encode(bytes const& _data)
...
+		size(_size),
+		blockSize(_blockSize)
+	{}
+
+	bytes hash = {};
+	size_t size = 0;
+	size_t blockSize = 0;
+};
+
+struct Link
```
Comment (chriseth) [inline libsolutil/IpfsHash.cpp]: I don't understand - in which way is the Link data structure different from the Chunk data structure? :)

Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: After removing links and lengths bytes, and also functions it ended up as the same structure :)

Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: I removed Link completely

File: libsolutil/IpfsHash.cpp (hunk 23 above)
Comment (mijovic) [reply libsolutil/IpfsHash.cpp]: I removed Link structure, so this is also fixed

**Review by chriseth — APPROVED** [review]

**Pull request merged**